google/cloud/aiplatform/source_config.py

```diff
@@ -0,0 +1,160 @@
+        
+    @property
+    def metadata(self) -> Dict:
+        return {"input_config": {"gcs_source": {"uri": self._source_uris}}}
+
+class GCSTabularSourceConfig(GCSSourceConfig):
+    @property
+    def metadata_schema_uri(self) -> str:
+        # Always return tabular as this config only supports tabular
+        return schema.dataset.metadata.tabular
+
+class GCSNonTabularSourceConfig(GCSSourceConfig, DataImportable):
+    def __init__(self, source_uris: [str], metadata_schema_uri: str, import_schema_uri: str, data_items_labels: Optional[Dict] = None):
+        """TODO
+
+        Args:
+            source_uris (Sequence[str]):
+                Required. Google Cloud Storage URI(-s) to the
+                input file(s). May contain wildcards. For more
+                information on wildcards, see
+                https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames.
+            metadata_schema_uri (str):
+                Required. Points to a YAML file stored on Google Cloud Storage
+                describing additional information about the Dataset. The schema
+                is defined as an OpenAPI 3.0.2 Schema Object. The schema files
+                that can be used here are found in gs://google-cloud-
+                aiplatform/schema/dataset/metadata/.
+            import_schema_uri (str):
+                Required. Points to a YAML file stored on Google Cloud
+                Storage describing the import format. Validation will be
+                done against the schema. The schema is defined as an
+                `OpenAPI 3.0.2 Schema
+                Object <https://tinyurl.com/y538mdwt>`__.
+            data_item_labels: (Optional[Dict]) = None
+                Labels that will be applied to newly imported DataItems. If
+                an identical DataItem as one being imported already exists
+                in the Dataset, then these labels will be appended to these
+                of the already existing one, and if labels with identical
+                key is imported before, the old label value will be
+                overwritten. If two DataItems are identical in the same
+                import data operation, the labels will be combined and if
+                key collision happens in this case, one of the values will
+                be picked randomly. Two DataItems are considered identical
+                if their content bytes are identical (e.g. image bytes or
+                pdf bytes). These labels will be overridden by Annotation
+                labels specified inside index file refenced by
+                [import_schema_uri][google.cloud.aiplatform.v1beta1.ImportDataConfig.import_schema_uri],
+                e.g. jsonl file.
+        """
+
+        if metadata_schema_uri == schema.dataset.metadata.tabular:
+            raise ValueError(
+                "Expected non-tabular metadata schema uri. For tabular metadata schema, use GCSTabularSourceConfig"
```

Comment:
Ew, the code smell is strong here. The cleaner way would be to create an interface to differentiate tabular schema types and non-tabular at the type level.

However, since Python won't enforce the types for us anyways, this code might be a better trade-off.